File: scala/src/main/scala/cookbook/RecipeThread.scala
package cookbook

import scala.concurrent.{Future, Await}
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration.Duration
import com.typesafe.scalalogging.Logger
import org.slf4j.LoggerFactory

class RecipeThread private extends Recipe {

    val logger = Logger(LoggerFactory getLogger this.getClass)

    def title() = {

        logger debug "# Thread"

    }

    def cook() = {

        val secs = List(1.0, 2.0, 3.0)

        var threads = List[Future[Unit]]()

        // スレッド開始

        for (sec <- secs) {
            threads = threads :+ Future { __sleep(sec) }
        }

        // スレッド終了待ち

        for (thread <- threads) {
            Await.ready(thread, Duration.Inf)
        }

        logger debug "all ended"

    }

    def __sleep(sec: Double) = {

        logger debug s"sleep(${sec}) started"
        Thread sleep (sec * 1000).toLong
        logger debug s"sleep(${sec}) ended"

    }

}

object RecipeThread {

    def apply() = {

        new RecipeThread

    }

    def main(args: Array[String]) {

        val recipe = RecipeThread.apply

        recipe.title
        recipe.cook

    }

}

File: java/src/main/java/cookbook/RecipeThread.java
package cookbook;

import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class RecipeThread implements Recipe {

    private static final Logger LOGGER = LoggerFactory.getLogger(RecipeThread.class);

    public void title() {

        LOGGER.debug("# Thread");

    }

    public void cook() {

        List<Double> secs = Arrays.asList(1.0, 2.0, 3.0);

        // スレッド生成

        List<Thread> threads = new ArrayList<>();
        for(double sec: secs) {
            Thread thread = new Thread(() -> {
                LOGGER.debug("sleep({}) started", sec);
                try {
                    Thread.sleep(Double.valueOf(sec*1000).longValue());
                    LOGGER.debug("sleep({}) ended", sec);
                } catch (InterruptedException ex) {
                    LOGGER.error(ex.getMessage(), ex);
                }
            });
            threads.add(thread);
        }

        // スレッド開始

        for(Thread thread: threads) {
            thread.start();
        }

        // スレッド終了待ち

        for(Thread thread: threads) {
            try {
                thread.join();
            } catch (InterruptedException ex) {
                LOGGER.error(ex.getMessage(), ex);
            }
        }

    }

    public static void main(String[] args) throws Exception {

        Recipe recipe = new RecipeThread();

        recipe.title();
        recipe.cook();

    }

}
